From 5744c7ca06b2f12bf60f7e242071e22abc97a2e7 Mon Sep 17 00:00:00 2001
From: Samuel Hym <samuel@tarides.com>
Date: Mon, 28 Apr 2025 17:36:04 +0200
Subject: [PATCH] Terminate the whole unikernel on calls to `exit`

In our use case, there will ever be only one "process", so terminate the
complete VM when it exits
So add a patch that routes musl's `_Exit` to `ukplat_{halt,crash}`
---
 ...the-whole-unikernel-on-calls-to-exit.patch | 32 +++++++++++++++++++
 1 file changed, 32 insertions(+)
 create mode 100644 patches/0022-Terminate-the-whole-unikernel-on-calls-to-exit.patch

diff --git a/patches/0022-Terminate-the-whole-unikernel-on-calls-to-exit.patch b/patches/0022-Terminate-the-whole-unikernel-on-calls-to-exit.patch
new file mode 100644
index 0000000..e881e2f
--- /dev/null
+++ b/patches/0022-Terminate-the-whole-unikernel-on-calls-to-exit.patch
@@ -0,0 +1,32 @@
+From 91ebc5bfcccfb5753e4df355322d241625d98f2b Mon Sep 17 00:00:00 2001
+From: Samuel Hym <samuel@tarides.com>
+Date: Mon, 28 Apr 2025 17:30:20 +0200
+Subject: [PATCH] Terminate the whole unikernel on calls to `exit`
+
+In our use case, there will ever be only one "process", so terminate the
+complete VM when it exits
+---
+ src/exit/_Exit.c | 9 +++++----
+ 1 file changed, 5 insertions(+), 4 deletions(-)
+
+diff --git a/src/exit/_Exit.c b/src/exit/_Exit.c
+index 7a6115c7..640b9e6e 100644
+--- a/src/exit/_Exit.c
++++ b/src/exit/_Exit.c
+@@ -1,8 +1,9 @@
+-#include <stdlib.h>
+-#include "syscall.h"
++#include "uk/plat/bootstrap.h"
+ 
+ _Noreturn void _Exit(int ec)
+ {
+-	__syscall(SYS_exit_group, ec);
+-	for (;;) __syscall(SYS_exit, ec);
++	if(ec == 0)
++		ukplat_halt();
++	else
++		ukplat_crash();
+ }
+-- 
+2.49.0
+
-- 
2.49.0

